### PR TITLE
feat: add self-upgrade feature

### PR DESCRIPTION
**Pull request by ankit-pn**
Feature is added as described in issue #31 .

### CODE REVIEW

File: cmd/commands/upgrade.go [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+            outFile.Close()
+        }
+    }
+}
+
+func upgradeCmd(ctx *cli.Context) error {
+	apiURL := "https://api.github.com/repos/version-fox/vfox/releases/latest"
+	releaseInfo, err := getLatestReleaseInfo(apiURL)
+	currVersion := fmt.Sprintf("v%s",ctx.App.Version);
+	latestVersion := releaseInfo.TagName;
```
Comment (aooohan) [inline cmd/commands/upgrade.go:160]: Indentation is inconsistent, you can use `go fmt` to format code.

File: cmd/commands/upgrade.go [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+    resp, err := http.Get(url)
+    if err != nil {
+        return err
+    }
+    defer resp.Body.Close()
+    _, err = io.Copy(out, resp.Body)
+    return err
+}
+
+func extractTarGz(gzipFilePath, destDir string) error {
```
Comment (aooohan) [inline cmd/commands/upgrade.go:110]: There is a decompression tool in the project. Please use that tool to replace this function. There is no need to reimplement it. 

> ⚠️ NOTE: Slightly different behavior

https://github.com/version-fox/vfox/blob/d9dd9937c979d0202beccf20cca4c4191bee68e5/internal/util/decompressor.go#L231

File: cmd/commands/upgrade.go [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+    if err := extractTarGz(tempFile, tempDir); err != nil {
+        return cli.Exit("Failed to extract file: "+err.Error(), 1)
+    }
+    folderName := strings.TrimSuffix(fileName, ".tar.gz")
+    newExePath := filepath.Join(tempDir,folderName, "vfox")
+    if err := os.Rename(newExePath, exePath); err != nil {
+        return cli.Exit("Failed to replace executable: "+err.Error(), 1)
+    }
+	if err := os.Chmod(exePath, 0755); err != nil {
+        panic("Failed to make executable: " + err.Error())
```
Comment (aooohan) [inline cmd/commands/upgrade.go:191]: Indentation is inconsistent, please use `go fmt` to format code.

File: cmd/commands/upgrade.go [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+	}
+
+	fileName := constructBinaryName(latestVersion)
+    exePath, err := os.Executable()
+    if err != nil {
+        return cli.Exit("Failed to get executable path: "+err.Error(), 1)
+    }
+
+    binURL, diffURL := getUrls(apiURL,currVersion)
+	exeDir := filepath.Dir(exePath)
```
Comment (aooohan) [inline cmd/commands/upgrade.go:175]: Indentation is inconsistent.

**Comment by ankit-pn** [comment]
I have made the changes as mentioned by you. Pls review it. Thanks

File: cmd/commands/upgrade.go [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+	tempDir := filepath.Join(exeDir, "update_vfox")
+
+	if err := downloadFile(tempFile, binURL); err != nil {
+		return cli.Exit("Failed to download file: "+err.Error(), 1)
+	}
+	if err := extractTarGz(tempFile, tempDir); err != nil {
+		return cli.Exit("Failed to extract file: "+err.Error(), 1)
+	}
+	newExePath := filepath.Join(tempDir, "vfox")
+	if err := os.Rename(newExePath, exePath); err != nil {
```
Comment (aooohan) [inline cmd/commands/upgrade.go:145]: If you directly replace it, it should be possible under `unix-like` system, but is it also possible under `windows`? Will it not prompt that the current file(`exePath`) is occupied? The current process has not ended yet.

File: cmd/commands/upgrade.go [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+	if err != nil {
+		return releaseInfo, err
+	}
+
+	err = json.Unmarshal(body, &releaseInfo)
+	return releaseInfo, err
+}
+
+func constructBinaryName(tagName string) string {
+	osType := strings.ToLower(runtime.GOOS)
```
Comment (baymax55) [inline cmd/commands/upgrade.go:61]: Does it mean that Windows system self-update is not supported?

File: cmd/commands/upgrade.go [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+	if osType == "darwin" {
```
Comment (baymax55) [inline cmd/commands/upgrade.go:62]: Does it mean that Windows system self-update is not supported?

File: cmd/commands/upgrade.go [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+		osType = "macos"
+	}
+
+	archType := runtime.GOARCH
+	if archType == "arm64" {
+		archType = "aarch64"
+	}
+
+	fileName := fmt.Sprintf("vfox_%s_%s_%s.tar.gz", tagName[1:], osType, archType)
```
Comment (baymax55) [inline cmd/commands/upgrade.go]: Does it mean that Windows system self-update is not supported?

File: cmd/commands/upgrade.go [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
```
Comment (baymax55) [inline cmd/commands/upgrade.go:71]: Does it mean that Windows system self-update is not supported?

**Comment by aooohan** [comment]
Close in favor of #239

**Pull request closed without merging**